// @orchetron/storm/dist/styles/defaults.js
export const DEFAULTS = {
    card: {
        borderStyle: "round",
        paddingLeft: 2,
        paddingRight: 2,
        paddingTop: 1,
        paddingBottom: 1,
    },
    alert: {
        borderStyle: "round",
        paddingLeft: 1,
        paddingRight: 1,
    },
    modal: {
        borderStyle: "round",
        padding: 1,
        width: 60,
        size: "md",
    },
    button: {
        paddingLeft: 1,
        paddingRight: 1,
    },
    table: {
        borderStyle: "single",
    },
    dataGrid: {
        borderStyle: "single",
    },
    toast: {
        duration: 3000,
    },
    accordion: {},
    collapsible: {},
    tabbedContent: {},
    form: {},
    confirmDialog: {
        borderStyle: "round",
        paddingX: 2,
        paddingY: 1,
    },
    richLog: {},
    listView: {},
    header: {
        borderStyle: "round",
    },
    footer: {
        borderStyle: "single",
    },
    progressBar: {
        width: 30,
        filledChar: "\u2588", // █
        emptyChar: "\u2591", // ░
    },
    spinner: {
        interval: 80,
    },
    divider: {
        width: 200,
    },
    separator: {
        width: 200,
    },
    scrollbar: {
        thumbChar: "\u2503", // ┃
        trackChar: "\u2502", // │
    },
};
