// @orchetron/storm/dist/components/core/RadioGroup.js
import React, { useCallback, useRef, createContext, useContext } from "react";
import { useInput } from "../../hooks/useInput.js";
import { useTui } from "../../context/TuiContext.js";
import { useColors } from "../../hooks/useColors.js";
import { usePluginProps } from "../../hooks/usePluginProps.js";
import { usePersonality } from "../../core/personality.js";
import { pickLayoutProps } from "../../styles/applyStyles.js";
import { FOCUS_CHARS } from "../../utils/focus-chars.js";
export const RadioGroupContext = createContext(null);
export function useRadioGroupContext() {
    const ctx = useContext(RadioGroupContext);
    if (!ctx)
        throw new Error("RadioGroup sub-components must be used inside RadioGroup.Root");
    return ctx;
}
function RadioGroupRoot({ value, onChange, highlightIndex = 0, onHighlightChange, children, }) {
    const colors = useColors();
    const { requestRender } = useTui();
    const onChangeRef = useRef(onChange);
    onChangeRef.current = onChange;
    const onHighlightRef = useRef(onHighlightChange);
    onHighlightRef.current = onHighlightChange;
    const ctx = {
        value,
        highlightIndex,
        select: (v) => { onChangeRef.current?.(v); requestRender(); },
        setHighlightIndex: (i) => { onHighlightRef.current?.(i); requestRender(); },
    };
    return React.createElement(RadioGroupContext.Provider, { value: ctx }, React.createElement("tui-box", { flexDirection: "column" }, children));
}
function RadioGroupCompoundOption({ option, index = 0, children }) {
    const colors = useColors();
    const { value, highlightIndex } = useRadioGroupContext();
    const isSelected = option.value === value;
    const isHighlighted = index === highlightIndex;
    const isDisabled = option.disabled === true;
    const indicator = isSelected ? "\u25CF" : "\u25CB";
    const indicatorColor = isDisabled ? colors.text.disabled : isSelected ? colors.brand.primary : colors.text.dim;
    if (children) {
        return React.createElement("tui-box", { flexDirection: "column" }, children);
    }
    const optionChildren = [];
    optionChildren.push(React.createElement("tui-box", { key: "row", flexDirection: "row" }, React.createElement("tui-text", { color: indicatorColor, ...(isDisabled ? { dim: true } : {}) }, ` ${indicator} `), React.createElement("tui-text", isDisabled ? { dim: true, color: colors.text.disabled } : isHighlighted ? { bold: true } : {}, option.label)));
    if (option.description !== undefined) {
        optionChildren.push(React.createElement("tui-text", { key: "desc", dim: true, color: colors.text.dim }, `    ${option.description}`));
    }
    return React.createElement("tui-box", { flexDirection: "column" }, ...optionChildren);
}
const FILLED = "\u25CF"; // ●
const EMPTY = "\u25CB"; // ○
const RadioGroupBase = React.memo(function RadioGroup(rawProps) {
    const colors = useColors();
    const personality = usePersonality();
    const props = usePluginProps("RadioGroup", rawProps);
    const { options, value, onChange, color = colors.brand.primary, bold: boldProp, dim: dimProp, direction = "column", isFocused = true, } = props;
    const { requestRender } = useTui();
    const selectedIdx = options.findIndex((o) => o.value === value);
    const highlightRef = useRef(selectedIdx >= 0 ? selectedIdx : 0);
    // Pure clamp
    if (highlightRef.current >= options.length) {
        highlightRef.current = Math.max(0, options.length - 1);
    }
    const effectiveIndex = highlightRef.current;
    const onChangeRef = useRef(onChange);
    onChangeRef.current = onChange;
    const handleInput = useCallback((event) => {
        // Helper: find next non-disabled index in a direction
        const findNext = (from, dir) => {
            let idx = from;
            for (let i = 0; i < options.length; i++) {
                idx = (idx + dir + options.length) % options.length;
                if (!options[idx]?.disabled)
                    return idx;
            }
            return from; // all disabled, stay put
        };
        if (event.key === "up" || (direction === "row" && event.key === "left")) {
            highlightRef.current = findNext(highlightRef.current, -1);
            requestRender();
        }
        else if (event.key === "down" ||
            (direction === "row" && event.key === "right")) {
            highlightRef.current = findNext(highlightRef.current, 1);
            requestRender();
        }
        else if (event.key === "return" || event.char === " ") {
            const opt = options[highlightRef.current];
            if (opt && !opt.disabled) {
                onChangeRef.current?.(opt.value);
            }
        }
        else if (event.char && event.char.length === 1 && /[a-zA-Z0-9]/.test(event.char)) {
            // Type-ahead: jump to first non-disabled option starting with this letter
            const ch = event.char.toLowerCase();
            const idx = options.findIndex((o) => !o.disabled && o.label.toLowerCase().startsWith(ch));
            if (idx >= 0) {
                highlightRef.current = idx;
                requestRender();
            }
        }
    }, [options, direction, requestRender]);
    useInput(handleInput, { isActive: isFocused && onChange !== undefined });
    const outerBoxProps = {
        role: "radiogroup",
        flexDirection: direction === "row" ? "row" : "column",
        "aria-label": props["aria-label"],
        ...pickLayoutProps(props),
    };
    return React.createElement("tui-box", outerBoxProps, ...options.map((option, index) => {
        const isSelected = option.value === value;
        const isHighlighted = index === effectiveIndex;
        const isDisabled = option.disabled === true;
        if (props.renderOption) {
            return React.createElement("tui-box", {
                key: option.value,
                flexDirection: "column",
                ...(direction === "row" ? { marginRight: 2 } : {}),
            }, props.renderOption(option, { isSelected, isHighlighted, isDisabled }));
        }
        const indicator = isSelected ? FILLED : EMPTY;
        const indicatorColor = isDisabled ? colors.text.disabled : isSelected ? color : colors.text.dim;
        const optionChildren = [];
        // Focus indicator prefix — style from personality
        const focusChar = FOCUS_CHARS[personality.interaction.focusIndicator] ?? "> ";
        const focusPrefix = isHighlighted && !isDisabled ? focusChar : " ".repeat(focusChar.length);
        // Main row: focus indicator + radio indicator + label
        optionChildren.push(React.createElement("tui-box", { key: "row", flexDirection: "row" }, React.createElement("tui-text", { color: isHighlighted ? colors.brand.primary : undefined, bold: isHighlighted && !isDisabled }, focusPrefix), React.createElement("tui-text", { color: indicatorColor, ...(isDisabled ? { dim: true } : {}) }, `${indicator} `), React.createElement("tui-text", isDisabled ? { dim: true, color: colors.text.disabled } : isHighlighted ? { bold: true } : {}, option.label)));
        // Description below label
        if (option.description !== undefined) {
            optionChildren.push(React.createElement("tui-text", { key: "desc", dim: true, color: colors.text.dim }, `    ${option.description}`));
        }
        return React.createElement("tui-box", {
            key: option.value,
            flexDirection: "column",
            ...(direction === "row" ? { marginRight: 2 } : {}),
        }, ...optionChildren);
    }));
});
export const RadioGroup = Object.assign(RadioGroupBase, {
    Root: RadioGroupRoot,
    Option: RadioGroupCompoundOption,
});
