// @orchetron/storm/dist/components/core/Tabs.js
import React, { useRef, useCallback, createContext, useContext } from "react";
import { useInput } from "../../hooks/useInput.js";
import { useTui } from "../../context/TuiContext.js";
import { useColors } from "../../hooks/useColors.js";
import { pickStyleProps } from "../../styles/applyStyles.js";
import { usePluginProps } from "../../hooks/usePluginProps.js";
import { usePersonality } from "../../core/personality.js";
import { findNextNavigable as findNextNav } from "../../utils/navigation.js";
function findNextTab(tabs, fromIndex, direction) {
    return findNextNav(tabs.length, fromIndex, direction, (i) => !tabs[i].disabled, -1);
}
export const TabsContext = createContext(null);
export function useTabsContext() {
    const ctx = useContext(TabsContext);
    if (!ctx)
        throw new Error("Tabs sub-components must be used inside Tabs.Root");
    return ctx;
}
function TabsRoot({ activeKey, onActiveKeyChange, orientation = "horizontal", children }) {
    const colors = useColors();
    const { requestRender } = useTui();
    const onChangeRef = useRef(onActiveKeyChange);
    onChangeRef.current = onActiveKeyChange;
    const ctx = {
        activeKey,
        setActiveKey: (key) => { onChangeRef.current?.(key); requestRender(); },
        orientation,
    };
    return React.createElement(TabsContext.Provider, { value: ctx }, children);
}
function TabsTrigger({ tabKey, disabled = false, closable = false, children, color: triggerColor }) {
    const colors = useColors();
    const { activeKey, setActiveKey } = useTabsContext();
    const isActive = activeKey === tabKey;
    const effectiveColor = triggerColor ?? colors.brand.primary;
    if (children) {
        return React.createElement("tui-box", { flexDirection: "row" }, children);
    }
    const textProps = {};
    if (disabled) {
        textProps["dim"] = true;
        textProps["color"] = colors.text.disabled;
    }
    else if (isActive) {
        textProps["bold"] = true;
        textProps["color"] = effectiveColor;
    }
    else {
        textProps["dim"] = true;
    }
    const labelText = closable ? `[ ${tabKey} \u00D7 ]` : `[ ${tabKey} ]`;
    return React.createElement("tui-text", textProps, labelText);
}
function TabsPanel({ tabKey, children }) {
    const { activeKey } = useTabsContext();
    if (activeKey !== tabKey)
        return null;
    return React.createElement("tui-box", { flexDirection: "column" }, children);
}
const TabsBase = React.memo(function Tabs(rawProps) {
    const colors = useColors();
    const personality = usePersonality();
    const props = usePluginProps("Tabs", rawProps);
    const { tabs, activeKey, onChange, onClose, color = colors.brand.primary, isFocused = true, orientation = "horizontal", } = props;
    const layoutProps = pickStyleProps(props);
    const onChangeRef = useRef(onChange);
    onChangeRef.current = onChange;
    const onCloseRef = useRef(onClose);
    onCloseRef.current = onClose;
    const tabsRef = useRef(tabs);
    tabsRef.current = tabs;
    const activeKeyRef = useRef(activeKey);
    activeKeyRef.current = activeKey;
    const handleInput = useCallback((event) => {
        const currentTabs = tabsRef.current;
        const cb = onChangeRef.current;
        if (!cb || currentTabs.length === 0)
            return;
        const currentIndex = currentTabs.findIndex((t) => t.key === activeKeyRef.current);
        const idx = currentIndex >= 0 ? currentIndex : 0;
        // Navigation keys depend on orientation
        const prevKey = orientation === "vertical" ? "up" : "left";
        const nextKey = orientation === "vertical" ? "down" : "right";
        if (event.key === prevKey) {
            const next = findNextTab(currentTabs, idx, -1);
            if (next >= 0)
                cb(currentTabs[next].key);
        }
        else if (event.key === nextKey) {
            const next = findNextTab(currentTabs, idx, 1);
            if (next >= 0)
                cb(currentTabs[next].key);
        }
        else if (event.char && /^[1-9]$/.test(event.char)) {
            const numIdx = parseInt(event.char, 10) - 1;
            if (numIdx < currentTabs.length && !currentTabs[numIdx].disabled) {
                cb(currentTabs[numIdx].key);
            }
        }
        else if ((event.key === "delete" || event.key === "backspace") && onCloseRef.current) {
            const activeTab = currentTabs[idx];
            if (activeTab && activeTab.closable) {
                onCloseRef.current(activeTab.key);
            }
        }
    }, [orientation]);
    useInput(handleInput, { isActive: isFocused !== false });
    const tabElements = tabs.map((tab) => {
        const isActive = tab.key === activeKey;
        const isDisabled = tab.disabled === true;
        if (props.renderTab) {
            return React.createElement("tui-box", { key: tab.key, flexDirection: "row" }, props.renderTab(tab, { isActive, isDisabled }));
        }
        const tabProps = { key: tab.key };
        if (isDisabled) {
            tabProps["dim"] = true;
            tabProps["color"] = colors.text.disabled;
        }
        else if (isActive) {
            tabProps["bold"] = true;
            tabProps["color"] = color;
        }
        else {
            tabProps["dim"] = true;
        }
        const labelText = tab.closable
            ? `[ ${tab.label} \u00D7 ]`
            : `[ ${tab.label} ]`;
        return React.createElement("tui-text", tabProps, labelText);
    });
    // Add separators between tabs
    const children = [];
    for (let i = 0; i < tabElements.length; i++) {
        children.push(tabElements[i]);
        if (i < tabElements.length - 1) {
            if (orientation === "vertical") {
                // No explicit separator needed in vertical — each tab is on its own row
            }
            else {
                children.push(React.createElement("tui-text", { key: `sep-${i}` }, " "));
            }
        }
    }
    const outerBoxProps = {
        role: "tablist",
        flexDirection: orientation === "vertical" ? "column" : "row",
        "aria-label": props["aria-label"],
        ...layoutProps,
    };
    return React.createElement("tui-box", outerBoxProps, ...children);
});
export const Tabs = Object.assign(TabsBase, {
    Root: TabsRoot,
    Trigger: TabsTrigger,
    Panel: TabsPanel,
});
