// @orchetron/storm/dist/components/extras/Footer.js
import React from "react";
import { useColors } from "../../hooks/useColors.js";
import { pickStyleProps } from "../../styles/applyStyles.js";
import { DEFAULTS } from "../../styles/defaults.js";
import { usePersonality } from "../../core/personality.js";
import { usePluginProps } from "../../hooks/usePluginProps.js";
const BORDER_CHARS = {
    single: "\u2500", // ─
    double: "\u2501", // ━
};
const DEFAULT_WIDTH = 80;
function renderBindings(bindings) {
    const colors = useColors();
    const children = [];
    for (let i = 0; i < bindings.length; i++) {
        const binding = bindings[i];
        if (i > 0) {
            children.push(React.createElement("tui-text", { key: `sep-${i}`, color: colors.text.dim }, " \u00B7 "));
        }
        children.push(React.createElement("tui-text", { key: `key-${i}`, color: colors.brand.primary, bold: true }, binding.key));
        children.push(React.createElement("tui-text", { key: `label-${i}`, color: colors.text.secondary }, ` ${binding.label}`));
    }
    return React.createElement("tui-box", { flexDirection: "row" }, ...children);
}
export const Footer = React.memo(function Footer(rawProps) {
    const colors = useColors();
    const props = usePluginProps("Footer", rawProps);
    const personality = usePersonality();
    const { children, borderStyle = DEFAULTS.footer.borderStyle, width, bindings, left, right, } = props;
    const layoutProps = pickStyleProps(props);
    const borderWidth = typeof width === "number" ? width : DEFAULT_WIDTH;
    const elements = [];
    if (borderStyle !== "none") {
        const borderChar = BORDER_CHARS[borderStyle] ?? BORDER_CHARS.single;
        const borderLine = borderChar.repeat(borderWidth);
        elements.push(React.createElement("tui-text", { key: "__border-top", color: colors.text.dim }, borderLine));
    }
    const hasLeftRight = left !== undefined || right !== undefined;
    if (bindings !== undefined && bindings.length > 0) {
        const bindingsContent = renderBindings(bindings);
        if (hasLeftRight) {
            // Combine left, bindings, and right
            const rowChildren = [];
            if (left !== undefined) {
                rowChildren.push(typeof left === "string"
                    ? React.createElement("tui-text", { key: "left", color: colors.text.secondary }, left)
                    : React.createElement("tui-box", { key: "left" }, left));
                rowChildren.push(React.createElement("tui-text", { key: "lsep" }, "  "));
            }
            rowChildren.push(React.createElement("tui-box", { key: "bindings" }, bindingsContent));
            if (right !== undefined) {
                rowChildren.push(React.createElement("tui-box", { key: "spacer", flex: 1 }));
                rowChildren.push(typeof right === "string"
                    ? React.createElement("tui-text", { key: "right", color: colors.text.secondary }, right)
                    : React.createElement("tui-box", { key: "right" }, right));
            }
            elements.push(React.createElement("tui-box", { key: "__content", flexDirection: "row", paddingLeft: 1 }, ...rowChildren));
        }
        else {
            elements.push(React.createElement("tui-box", { key: "__content", paddingLeft: 1 }, bindingsContent));
        }
    }
    else if (hasLeftRight) {
        // left/right layout without bindings
        const rowChildren = [];
        if (left !== undefined) {
            rowChildren.push(typeof left === "string"
                ? React.createElement("tui-text", { key: "left", color: colors.text.secondary }, left)
                : React.createElement("tui-box", { key: "left" }, left));
        }
        rowChildren.push(React.createElement("tui-box", { key: "spacer", flex: 1 }));
        if (right !== undefined) {
            rowChildren.push(typeof right === "string"
                ? React.createElement("tui-text", { key: "right", color: colors.text.secondary }, right)
                : React.createElement("tui-box", { key: "right" }, right));
        }
        elements.push(React.createElement("tui-box", { key: "__content", flexDirection: "row", paddingLeft: 1 }, ...rowChildren));
    }
    else if (props.renderContent) {
        elements.push(React.createElement("tui-box", { key: "__content", paddingLeft: 1 }, props.renderContent(children)));
    }
    else if (children !== undefined) {
        elements.push(React.createElement("tui-box", { key: "__content", paddingLeft: 1 }, children));
    }
    const outerBoxProps = {
        flexDirection: "column",
        ...layoutProps,
    };
    return React.createElement("tui-box", outerBoxProps, ...elements);
});
