// @orchetron/storm/dist/components/extras/Stepper.js
import React, { createContext, useContext } from "react";
import { useColors } from "../../hooks/useColors.js";
import { pickStyleProps } from "../../styles/applyStyles.js";
import { usePluginProps } from "../../hooks/usePluginProps.js";
export const StepperContext = createContext(null);
export function useStepperContext() {
    const ctx = useContext(StepperContext);
    if (!ctx)
        throw new Error("Stepper sub-components must be used inside Stepper.Root");
    return ctx;
}
function StepperRoot({ activeStep, orientation = "horizontal", completedColor: completedColorProp, activeColor: activeColorProp, pendingColor: pendingColorProp, children, }) {
    const colors = useColors();
    const completedColor = completedColorProp ?? colors.success;
    const activeColor = activeColorProp ?? colors.brand.primary;
    const pendingColor = pendingColorProp ?? colors.text.dim;
    const counterRef = { current: 0 };
    const ctx = {
        activeStep,
        orientation,
        completedColor,
        activeColor,
        pendingColor,
        stepCount: React.Children.count(children),
        registerStep: () => counterRef.current++,
    };
    return React.createElement(StepperContext.Provider, { value: ctx }, React.createElement("tui-box", { flexDirection: orientation === "vertical" ? "column" : "row" }, children));
}
function StepperStep({ index, label, description, children }) {
    const colors = useColors();
    const { activeStep, orientation, completedColor, activeColor, pendingColor } = useStepperContext();
    const status = getStatus(index, activeStep);
    const stepColor = getColor(status, completedColor, activeColor, pendingColor);
    if (children) {
        return React.createElement("tui-box", { flexDirection: orientation === "vertical" ? "column" : "row" }, children);
    }
    if (orientation === "vertical") {
        const indicator = getIndicator(status);
        const elements = [];
        elements.push(React.createElement("tui-box", { key: "step", flexDirection: "row" }, React.createElement("tui-text", { key: "ind", color: stepColor, bold: status === "active" }, `${indicator} `), React.createElement("tui-text", { key: "label", color: stepColor, bold: status === "active" }, label)));
        if (description) {
            elements.push(React.createElement("tui-text", { key: "desc", color: status === "active" ? colors.text.secondary : colors.text.dim }, `  ${description}`));
        }
        return React.createElement("tui-box", { flexDirection: "column" }, ...elements);
    }
    // Horizontal
    return React.createElement("tui-text", { color: stepColor, bold: status === "active" }, `${circledNumber(index + 1)} ${label}`);
}
function getStatus(index, activeStep) {
    if (index < activeStep)
        return "completed";
    if (index === activeStep)
        return "active";
    return "pending";
}
function getColor(status, completedColor, activeColor, pendingColor) {
    if (status === "completed")
        return completedColor;
    if (status === "active")
        return activeColor;
    return pendingColor;
}
function getIndicator(status) {
    if (status === "completed")
        return "\u2713"; // check mark
    if (status === "active")
        return "\u25CF"; // filled circle
    return "\u25CB"; // empty circle
}
// Circled number characters for 1-20
const CIRCLED_NUMBERS = [
    "\u2460", "\u2461", "\u2462", "\u2463", "\u2464",
    "\u2465", "\u2466", "\u2467", "\u2468", "\u2469",
    "\u246A", "\u246B", "\u246C", "\u246D", "\u246E",
    "\u246F", "\u2470", "\u2471", "\u2472", "\u2473",
];
function circledNumber(n) {
    if (n >= 1 && n <= 20)
        return CIRCLED_NUMBERS[n - 1];
    return `(${n})`;
}
function renderHorizontal(steps, activeStep, completedColor, activeColor, pendingColor, colors) {
    const elements = [];
    for (let i = 0; i < steps.length; i++) {
        const step = steps[i];
        const status = getStatus(i, activeStep);
        const stepColor = getColor(status, completedColor, activeColor, pendingColor);
        // Step circle + label
        elements.push(React.createElement("tui-text", {
            key: `step-${i}`,
            color: stepColor,
            bold: status === "active",
        }, `${circledNumber(i + 1)} ${step.label}`));
        // Connector line (not after last step)
        if (i < steps.length - 1) {
            const connectorColor = i < activeStep ? completedColor : pendingColor;
            elements.push(React.createElement("tui-text", { key: `conn-${i}`, color: connectorColor }, " \u2500\u2500\u2500 "));
        }
    }
    return React.createElement("tui-box", { flexDirection: "row" }, ...elements);
}
function renderVertical(steps, activeStep, completedColor, activeColor, pendingColor, colors) {
    const elements = [];
    for (let i = 0; i < steps.length; i++) {
        const step = steps[i];
        const status = getStatus(i, activeStep);
        const stepColor = getColor(status, completedColor, activeColor, pendingColor);
        const indicator = getIndicator(status);
        // Step row: indicator + label
        const stepChildren = [];
        stepChildren.push(React.createElement("tui-text", { key: "ind", color: stepColor, bold: status === "active" }, `${indicator} `));
        stepChildren.push(React.createElement("tui-text", { key: "label", color: stepColor, bold: status === "active" }, step.label));
        elements.push(React.createElement("tui-box", { key: `step-${i}`, flexDirection: "row" }, ...stepChildren));
        // Description (if present)
        if (step.description) {
            elements.push(React.createElement("tui-text", {
                key: `desc-${i}`,
                color: status === "active" ? colors.text.secondary : colors.text.dim,
            }, `  ${step.description}`));
        }
        // Vertical connector line (not after last step)
        if (i < steps.length - 1) {
            const lineColor = i < activeStep ? completedColor : pendingColor;
            elements.push(React.createElement("tui-text", { key: `line-${i}`, color: lineColor }, "\u2502"));
        }
    }
    return React.createElement("tui-box", { flexDirection: "column" }, ...elements);
}
const StepperBase = React.memo(function Stepper(rawProps) {
    const colors = useColors();
    const props = usePluginProps("Stepper", rawProps);
    const { steps, activeStep: rawActiveStep, orientation = "horizontal", completedColor = colors.success, activeColor = colors.brand.primary, pendingColor = colors.text.dim, } = props;
    const layoutProps = pickStyleProps(props);
    // Allow activeStep >= steps.length to indicate all completed
    const activeStep = steps.length > 0
        ? Math.max(0, rawActiveStep)
        : 0;
    const outerBoxProps = {
        flexDirection: "column",
        role: "group",
        ...layoutProps,
    };
    // Custom render delegate for each step
    if (props.renderStep) {
        const stepElements = steps.map((step, i) => {
            const status = getStatus(i, activeStep);
            return React.createElement(React.Fragment, { key: `step-${i}` }, props.renderStep(step, { isActive: status === "active", isCompleted: status === "completed", index: i }));
        });
        const customInner = React.createElement("tui-box", { flexDirection: orientation === "vertical" ? "column" : "row" }, ...stepElements);
        if (Object.keys(layoutProps).length === 0)
            return customInner;
        return React.createElement("tui-box", outerBoxProps, customInner);
    }
    const inner = orientation === "vertical"
        ? renderVertical(steps, activeStep, completedColor, activeColor, pendingColor, colors)
        : renderHorizontal(steps, activeStep, completedColor, activeColor, pendingColor, colors);
    // If no layout style props provided, return inner directly to avoid an extra wrapper
    if (Object.keys(layoutProps).length === 0) {
        return inner;
    }
    return React.createElement("tui-box", outerBoxProps, inner);
});
export const Stepper = Object.assign(StepperBase, {
    Root: StepperRoot,
    Step: StepperStep,
});
