// @orchetron/storm/dist/utils/table-render.js
import { COL_WIDTH_SAMPLE_SIZE, padCell } from "./format.js";
/**
 * Compute column widths by sampling data rows.
 * If a column has an explicit `width`, it is used as-is.
 * Otherwise the width is the max of the header length and the longest
 * value in the first COL_WIDTH_SAMPLE_SIZE rows.
 *
 * @param headerField  Which field on the column object holds the header text
 *                     ("header" for Table, "label" for DataGrid).
 */
export function computeColumnWidths(columns, data, headerField = "header", extraPadding = 0) {
    const sampleCount = Math.min(data.length, COL_WIDTH_SAMPLE_SIZE);
    return columns.map((col) => {
        if (col.width !== undefined)
            return col.width;
        const headerText = String(col[headerField] ?? "");
        let max = headerText.length + extraPadding;
        for (let i = 0; i < sampleCount; i++) {
            const row = data[i];
            const val = row[col.key];
            const len = val !== undefined ? String(val).length : 0;
            if (len > max)
                max = len;
        }
        return max;
    });
}
// ─── Separator / divider line ───────────────────────────────────────────
/**
 * Build a horizontal separator string that sits between header and body.
 * Each column segment is `─` repeated (colWidth + 2) for the 1-char padding
 * on each side, joined by the given `joiner` character.
 *
 * Table uses "┼" (`\u253C`), DataGrid uses "─" (`\u2500`).
 */
export function buildSeparatorText(colWidths, joiner) {
    return colWidths
        .map((w) => "\u2500".repeat(w + 2))
        .join(joiner);
}
// ─── Zebra stripe background ────────────────────────────────────────────
/**
 * Determine whether a row should receive the zebra-stripe background.
 * Returns `true` when striping is enabled, the row index is odd,
 * and the row is not in a highlighted/selected state.
 */
export function shouldStripe(rowIndex, stripeEnabled, isHighlighted) {
    return stripeEnabled && rowIndex % 2 === 1 && !isHighlighted;
}
/**
 * Compute the visible row window for virtualized rendering.
 */
export function computeVirtualWindow(totalRows, maxVisibleRows, scrollOffset) {
    const needsVirtualization = totalRows > maxVisibleRows;
    const clampedOffset = needsVirtualization
        ? Math.max(0, Math.min(scrollOffset, totalRows - maxVisibleRows))
        : 0;
    const start = clampedOffset;
    const end = needsVirtualization
        ? Math.min(clampedOffset + maxVisibleRows, totalRows)
        : totalRows;
    return { start, end, total: totalRows, needsVirtualization };
}
// ─── Row count indicator text ───────────────────────────────────────────
/**
 * Format the "Showing X-Y of Z" (Table) or "X-Y of Z" (DataGrid) indicator.
 *
 * @param prefix  Text before the range, e.g. "Showing " or "" (empty).
 */
export function formatRowIndicator(start, end, total, prefix = "") {
    return `${prefix}${start + 1}-${end} of ${total.toLocaleString()}`;
}
// ─── Header text with sort indicator ────────────────────────────────────
/**
 * Append a sort direction arrow (▲ or ▼) to a header label when the
 * column is the active sort column.
 */
export function headerTextWithSort(label, columnKey, sortColumn, sortDirection) {
    if (sortColumn === columnKey && sortDirection) {
        return label + (sortDirection === "asc" ? " \u25B2" : " \u25BC");
    }
    return label;
}
// ─── Padded row line builder ────────────────────────────────────────────
/**
 * Build a single row line string: each cell is ` <padded> ` joined by the
 * given separator character (typically "│").
 */
export function buildRowLine(columns, row, colWidths, separator = "\u2502") {
    return columns.map((col, ci) => {
        const val = row[col.key];
        const text = val !== undefined ? String(val) : "";
        return " " + padCell(text, colWidths[ci], col.align ?? "left") + " ";
    }).join(separator);
}
