// @orchetron/storm/dist/input/mouse.js
/**
 * Mouse event parsing — SGR (1006) and X11 (1002) protocols.
 */
const SGR_REGEX = /^\x1b\[<(\d+);(\d+);(\d+)([mM])/;
const X11_PREFIX = "\x1b[M";
/**
 * Try to parse a mouse event from the beginning of a buffer.
 * Returns the parsed event and how many bytes were consumed,
 * or null if the buffer doesn't start with a mouse sequence.
 */
export function parseMouseEvent(buffer) {
    return parseSGR(buffer) ?? parseX11(buffer);
}
function parseSGR(buffer) {
    const match = SGR_REGEX.exec(buffer);
    if (!match)
        return null;
    const cb = Number(match[1]);
    const cx = Math.max(0, Math.min(Number(match[2]) - 1, 10000)); // 1-indexed → 0-indexed, clamped
    const cy = Math.max(0, Math.min(Number(match[3]) - 1, 10000));
    const release = match[4] === "m";
    const { button, action: baseAction } = decodeButton(cb);
    const action = release ? "release" : baseAction;
    return {
        event: {
            button,
            action,
            x: cx,
            y: cy,
            shift: (cb & 4) !== 0,
            meta: (cb & 8) !== 0,
            ctrl: (cb & 16) !== 0,
            raw: match[0],
        },
        consumed: match[0].length,
    };
}
function parseX11(buffer) {
    if (!buffer.startsWith(X11_PREFIX))
        return null;
    if (buffer.length < X11_PREFIX.length + 3)
        return null;
    const offset = X11_PREFIX.length;
    const cb = buffer.charCodeAt(offset) - 32;
    const cx = Math.max(0, Math.min(buffer.charCodeAt(offset + 1) - 33, 10000)); // 1-indexed → 0-indexed, clamped
    const cy = Math.max(0, Math.min(buffer.charCodeAt(offset + 2) - 33, 10000));
    const consumed = offset + 3;
    const { button, action } = decodeButton(cb);
    return {
        event: {
            button,
            action,
            x: cx,
            y: cy,
            shift: (cb & 4) !== 0,
            meta: (cb & 8) !== 0,
            ctrl: (cb & 16) !== 0,
            raw: buffer.slice(0, consumed),
        },
        consumed,
    };
}
function decodeButton(cb) {
    const low2 = cb & 3;
    const isMotion = (cb & 32) !== 0;
    const isScroll = (cb & 64) !== 0;
    if (isScroll) {
        // Scroll wheel
        if (low2 === 0)
            return { button: "scroll-up", action: "press" };
        if (low2 === 1)
            return { button: "scroll-down", action: "press" };
        if (low2 === 2)
            return { button: "scroll-left", action: "press" };
        return { button: "scroll-right", action: "press" };
    }
    if (isMotion) {
        // Motion event
        const motionButton = low2 === 0 ? "left" : low2 === 1 ? "middle" : low2 === 2 ? "right" : "none";
        return { button: motionButton, action: "move" };
    }
    // Button press
    if (low2 === 0)
        return { button: "left", action: "press" };
    if (low2 === 1)
        return { button: "middle", action: "press" };
    if (low2 === 2)
        return { button: "right", action: "press" };
    return { button: "none", action: "release" }; // low2 === 3 = release in X11
}
/**
 * Check if the buffer starts with an incomplete mouse sequence.
 * Used to hold the buffer and wait for more data.
 */
export function isIncompleteMouseSequence(buffer) {
    // SGR prefix: \x1b[< — could be start of SGR mouse event
    if (buffer === "\x1b" || buffer === "\x1b[" || buffer === "\x1b[<")
        return true;
    if (buffer.startsWith("\x1b[<")) {
        // We have the prefix but no terminating m or M yet
        return buffer.length < 50 && !/[mM]/.test(buffer.slice(3));
    }
    // X11 prefix: \x1b[M — need 3 more bytes
    if (buffer.startsWith(X11_PREFIX)) {
        return buffer.length < X11_PREFIX.length + 3;
    }
    return false;
}
