// @orchetron/storm/dist/index.js
/** @packageDocumentation */
export { EMPTY_CELL, DEFAULT_COLOR, Attr, rgb, parseColor, cellEquals, makeCell, styleToAttrs, styleToCellProps, BORDER_CHARS, } from "./core/types.js";
export { ScreenBuffer, WIDE_CHAR_PLACEHOLDER } from "./core/buffer.js";
export { DiffRenderer, isWasmAccelerated } from "./core/diff.js";
export { Screen } from "./core/screen.js";
export { ease, spring } from "./core/easing.js";
export { detectTerminal, terminalInfo, } from "./core/terminal-detect.js";
export { detectImageCaps, bestImageProtocolDetailed, } from "./core/terminal-caps.js";
export { createAdaptiveConfig, bestImageProtocol, bestKeyboardProtocol, bestColorDepth, enableKittyKeyboard, enableSyncOutput, adaptiveChar, adaptiveBorder, } from "./core/adaptive.js";
export { RenderErrorBoundary, } from "./core/error-boundary.js";
export { OutputBuffer, } from "./core/backpressure.js";
export { MAX_LAYOUT_DEPTH, MAX_CHILDREN, MAX_BUFFER_WIDTH, MAX_BUFFER_HEIGHT, clampDimension, validateLayoutProps, isTerminalAlive, } from "./core/guards.js";
export { ESC, CSI, CURSOR_HIDE, CURSOR_SHOW, CURSOR_SAVE, CURSOR_RESTORE, CLEAR_SCREEN, CLEAR_LINE, ALT_SCREEN_ENTER, ALT_SCREEN_EXIT, MOUSE_ENABLE, MOUSE_DISABLE, RESET, cursorTo, fgColor, bgColor, fullSgr, diffSgr, setScrollRegion, RESET_SCROLL_REGION, setColorDepth, getColorDepth, rgbTo256, rgbTo16, } from "./core/ansi.js";
export { parseMouseEvent, isIncompleteMouseSequence } from "./input/mouse.js";
export { parseKeys } from "./input/keyboard.js";
export { InputManager } from "./input/manager.js";
export { computeLayout, measureNaturalHeight, measureNaturalWidth, } from "./layout/engine.js";
export { render } from "./reconciler/render.js";
export { renderToString, } from "./reconciler/render-to-string.js";
export { Box } from "./components/core/Box.js";
export { Text } from "./components/core/Text.js";
export { ScrollView } from "./components/core/ScrollView.js";
export { TextInput } from "./components/core/TextInput.js";
export { Spinner } from "./components/core/Spinner.js";
export { Spacer } from "./components/core/Spacer.js";
export { Newline } from "./components/core/Newline.js";
export { Link } from "./components/extras/Link.js";
export { Static } from "./components/core/Static.js";
export { SelectInput } from "./components/core/SelectInput.js";
export { Overlay } from "./components/core/Overlay.js";
export { Divider } from "./components/core/Divider.js";
export { Table } from "./components/data/Table.js";
export { ProgressBar } from "./components/core/ProgressBar.js";
export { Tabs } from "./components/core/Tabs.js";
export { Tree } from "./components/data/Tree.js";
export { Toast, ToastContainer } from "./components/extras/Toast.js";
export { Checkbox } from "./components/core/Checkbox.js";
export { RadioGroup } from "./components/core/RadioGroup.js";
export { Switch } from "./components/core/Switch.js";
export { ChatInput } from "./components/extras/ChatInput.js";
export { TextArea } from "./components/core/TextArea.js";
export { Button } from "./components/core/Button.js";
export { ListView } from "./components/core/ListView.js";
export { Collapsible } from "./components/extras/Collapsible.js";
export { Sparkline } from "./components/data/Sparkline.js";
export { Header } from "./components/extras/Header.js";
export { Footer } from "./components/extras/Footer.js";
export { RichLog } from "./components/data/RichLog.js";
export { Paginator } from "./components/extras/Paginator.js";
export { Timer } from "./components/extras/Timer.js";
export { FilePicker } from "./components/extras/FilePicker.js";
export { VirtualList } from "./components/core/VirtualList.js";
export { Gauge } from "./components/data/Gauge.js";
export { GradientProgress } from "./components/effects/GradientProgress.js";
export { RevealTransition } from "./components/effects/RevealTransition.js";
export { Image, isColoredUnderlineSupported } from "./components/effects/Image.js";
export { Shadow } from "./components/effects/Shadow.js";
export { Card } from "./components/extras/Card.js";
export { GlowText } from "./components/effects/GlowText.js";
export { GradientBorder } from "./components/effects/GradientBorder.js";
export { Kbd } from "./components/extras/Kbd.js";
export { LineChart } from "./components/data/LineChart.js";
export { AreaChart } from "./components/data/AreaChart.js";
export { ScatterPlot } from "./components/data/ScatterPlot.js";
export { Heatmap } from "./components/data/Heatmap.js";
export { Histogram } from "./components/data/Histogram.js";
export { BarChart } from "./components/data/BarChart.js";
export { Heading } from "./components/extras/Heading.js";
export { Paragraph } from "./components/extras/Paragraph.js";
export { Diagram } from "./components/effects/Diagram.js";
export { Canvas } from "./components/effects/canvas/index.js";
export { SearchInput } from "./components/extras/SearchInput.js";
export { MaskedInput } from "./components/extras/MaskedInput.js";
export { Form } from "./components/extras/Form.js";
export { DataGrid } from "./components/data/DataGrid.js";
export { OrderedList } from "./components/extras/OrderedList.js";
export { UnorderedList } from "./components/extras/UnorderedList.js";
export { DefinitionList } from "./components/data/DefinitionList.js";
export { Pretty } from "./components/data/Pretty.js";
export { Gradient } from "./components/effects/Gradient.js";
export { Digits } from "./components/effects/Digits.js";
export { Placeholder } from "./components/extras/Placeholder.js";
export { Accordion } from "./components/extras/Accordion.js";
export { Alert } from "./components/extras/Alert.js";
export { Avatar } from "./components/extras/Avatar.js";
export { Badge } from "./components/extras/Badge.js";
export { Breadcrumb } from "./components/extras/Breadcrumb.js";
export { Calendar } from "./components/extras/Calendar.js";
export { ConfirmDialog } from "./components/extras/ConfirmDialog.js";
export { ContentSwitcher } from "./components/extras/ContentSwitcher.js";
export { DirectoryTree } from "./components/extras/DirectoryTree.js";
export { KeyboardHelp } from "./components/extras/KeyboardHelp.js";
export { HelpPanel } from "./components/extras/HelpPanel.js";
export { Menu } from "./components/extras/Menu.js";
export { OptionList } from "./components/extras/OptionList.js";
export { Modal } from "./components/core/Modal.js";
export { Select } from "./components/core/Select.js";
export { SelectionList } from "./components/extras/SelectionList.js";
export { StatusMessage } from "./components/extras/StatusMessage.js";
export { Stepper } from "./components/extras/Stepper.js";
export { Stopwatch } from "./components/extras/Stopwatch.js";
export { TabbedContent } from "./components/extras/TabbedContent.js";
export { Tag } from "./components/extras/Tag.js";
export { Tooltip } from "./components/extras/Tooltip.js";
export { FocusGroup } from "./components/core/FocusGroup.js";
export { DiffView } from "./components/extras/DiffView.js";
export { InlineDiff } from "./components/extras/InlineDiff.js";
export { ErrorBoundary } from "./components/core/ErrorBoundary.js";
export { Transition } from "./components/effects/Transition.js";
export { AnimatePresence } from "./components/effects/AnimatePresence.js";
export { Markdown } from "./components/extras/Markdown.js";
export { MarkdownViewer } from "./components/extras/MarkdownViewer.js";
export { CommandPalette } from "./components/extras/CommandPalette.js";
export { DatePicker } from "./components/extras/DatePicker.js";
export { LoadingIndicator } from "./components/extras/LoadingIndicator.js";
export { Welcome } from "./components/extras/Welcome.js";
export { OperationTree, StreamingText, SyntaxHighlight, registerLanguage, getLanguage, getSupportedLanguages, ShimmerText, BlinkDot, ApprovalPrompt, CommandDropdown, StatusLine, MessageBubble, PerformanceHUD, TokenStream, ContextWindow, CostTracker, ModelBadge, CommandBlock, } from "./widgets/index.js";
export { useColors } from "./hooks/useColors.js";
export { useCleanup } from "./hooks/useCleanup.js";
export { useAsyncCleanup } from "./hooks/useAsyncCleanup.js";
export { useForceUpdate } from "./hooks/useForceUpdate.js";
export { useInput } from "./hooks/useInput.js";
export { useMouse } from "./hooks/useMouse.js";
export { useTerminal } from "./hooks/useTerminal.js";
export { useFocus } from "./hooks/useFocus.js";
export { useScroll } from "./hooks/useScroll.js";
export { useMeasure } from "./hooks/useMeasure.js";
export { useApp } from "./hooks/useApp.js";
export { useStdin } from "./hooks/useStdin.js";
export { useStdout } from "./hooks/useStdout.js";
export { useStderr } from "./hooks/useStderr.js";
export { useFocusManager } from "./hooks/useFocusManager.js";
export { useAnimation } from "./hooks/useAnimation.js";
export { usePaste } from "./hooks/usePaste.js";
export { useInterval } from "./hooks/useInterval.js";
export { useTimeout } from "./hooks/useTimeout.js";
export { useVirtualList } from "./hooks/useVirtualList.js";
export { useClipboard } from "./hooks/useClipboard.js";
export { useAccessibility } from "./hooks/useAccessibility.js";
export { useReducedMotion } from "./hooks/useReducedMotion.js";
export { useAdaptive } from "./hooks/useAdaptive.js";
export { usePluginManager } from "./hooks/usePlugin.js";
export { usePluginProps } from "./hooks/usePluginProps.js";
export { useBuffer } from "./hooks/useBuffer.js";
export { useAnnounce } from "./hooks/useAnnounce.js";
export { useTransition } from "./hooks/useTransition.js";
export { useTextCycler } from "./hooks/useTextCycler.js";
export { useCollapsibleContent } from "./hooks/useCollapsibleContent.js";
export { useModeCycler } from "./hooks/useModeCycler.js";
export { useGhostText } from "./hooks/useGhostText.js";
export { useInlinePrompt } from "./hooks/useInlinePrompt.js";
export { useEasedInterval } from "./hooks/useEasedInterval.js";
export { useTypeahead } from "./hooks/useTypeahead.js";
export { useSearchFilter } from "./hooks/useSearchFilter.js";
export { useMultiSelect } from "./hooks/useMultiSelect.js";
export { useUndoRedo } from "./hooks/useUndoRedo.js";
export { useCommandPalette } from "./hooks/useCommandPalette.js";
export { useHotkey } from "./hooks/useHotkey.js";
export { useTimer } from "./hooks/useTimer.js";
export { useConfirmAction } from "./hooks/useConfirmAction.js";
export { useNotification } from "./hooks/useNotification.js";
export { useWizard } from "./hooks/useWizard.js";
export { useStreamConsumer } from "./hooks/useStreamConsumer.js";
export { useBatchAction } from "./hooks/useBatchAction.js";
export { useKeyChord } from "./hooks/useKeyChord.js";
export { useSortable } from "./hooks/useSortable.js";
export { useHistory } from "./hooks/useHistory.js";
export { useAsyncLoader } from "./hooks/useAsyncLoader.js";
export { usePersistentState, memoryStorage } from "./hooks/usePersistentState.js";
export { useContextMenu } from "./hooks/useContextMenu.js";
export { useDragReorder } from "./hooks/useDragReorder.js";
export { useInfiniteScroll } from "./hooks/useInfiniteScroll.js";
export { useCopyPasteBuffer } from "./hooks/useCopyPasteBuffer.js";
export { useLocale } from "./hooks/useLocale.js";
export { paintBackgroundToBuffer } from "./reconciler/renderer.js";
export { useSelectBehavior, useListBehavior, useMenuBehavior, useTreeBehavior, useTabsBehavior, useAccordionBehavior, usePaginatorBehavior, useStepperBehavior, useTableBehavior, useVirtualListBehavior, useDialogBehavior, useToastBehavior, useFormBehavior, useCalendarBehavior, useCollapsibleBehavior, } from "./hooks/headless/index.js";
export { detectAccessibility, meetsContrast, relativeLuminance, contrastRatio, announce, } from "./core/accessibility.js";
export { ariaToAnnouncement, describeButton, describeCheckbox, describeTextInput, describeProgressBar, describeTab, describeMenuItem, describeTreeItem, describeListItem, describeDialog, describeAlert, } from "./core/aria.js";
export { getBoundingBox, getWidth, getHeight, hitTest, } from "./core/measurement.js";
export { EN, PLURAL_EN, PLURAL_AR, PLURAL_FR, PLURAL_RU, PLURAL_JA, registerLocale, getLocale, getRegisteredLocales, formatNumber, t, plural, LocaleContext, LocaleProvider, } from "./core/i18n.js";
export { useTui, TuiContext } from "./context/TuiContext.js";
export { colors, useTheme, ThemeProvider, ThemeContext, extendTheme, createTheme, extractThemeOverrides, spacing, arcticTheme, midnightTheme, emberTheme, mistTheme, voltageTheme, duskTheme, horizonTheme, neonTheme, calmTheme, highContrastTheme, monochromeTheme, loadTheme, parseTheme, saveTheme, serializeTheme, validateTheme, validateContrast, generateShades, generateThemeShades, } from "./theme/index.js";
export { defaultPersonality, createPersonality, mergePersonality, PersonalityProvider, usePersonality, PersonalityContext, } from "./core/personality.js";
export { defaultPreset, minimalPreset, hackerPreset, playfulPreset, } from "./core/personality-presets.js";
export { mergeBoxStyles, pickStyleProps, pickLayoutProps, DEFAULTS } from "./styles/index.js";
export { createStyleSheet, StyleSheet } from "./core/stylesheet.js";
export { StyleProvider, StyleContext, useStyles } from "./core/style-provider.js";
export { parseStormCSS, createStyleSheetLoader, } from "./core/stylesheet-loader.js";
export { useStyleSheet } from "./hooks/useStyleSheet.js";
export { FocusManager } from "./core/focus.js";
export { RenderContext } from "./core/render-context.js";
export { AnimationScheduler } from "./core/animation-scheduler.js";
export { PluginManager, PluginBus, } from "./core/plugin.js";
export { vimModePlugin } from "./plugins/vim-mode.js";
export { compactModePlugin } from "./plugins/compact-mode.js";
export { autoScrollPlugin } from "./plugins/auto-scroll.js";
export { screenshotPlugin } from "./plugins/screenshot.js";
export { statusBarPlugin } from "./plugins/status-bar.js";
export { MiddlewarePipeline, scanlineMiddleware, fpsCounterMiddleware, debugBorderMiddleware, darkenColor, } from "./core/middleware.js";
export { createInspectorMiddleware, } from "./devtools/inspector.js";
export { serializeTree } from "./devtools/tree-view.js";
export { createPerformanceMonitor, } from "./devtools/performance-monitor.js";
export { createEventLogger, } from "./devtools/event-logger.js";
export { createDevToolsOverlay, } from "./devtools/devtools-overlay.js";
export { createTimeTravel, } from "./devtools/time-travel.js";
export { createRenderHeatmap, } from "./devtools/render-heatmap.js";
export { createAccessibilityAudit, } from "./devtools/accessibility-audit.js";
export { enableDevTools, } from "./devtools/enable.js";
export { createProfiler, } from "./devtools/profiler.js";
export { setActiveProfiler, getActiveProfiler, } from "./devtools/profiler-registry.js";
export { enableCrashLog, } from "./devtools/crash-log.js";
export { BrailleCanvas, BRAILLE_BASE } from "./utils/braille-canvas.js";
export { createAnimation, tickAnimation, easings, } from "./utils/animate.js";
export { enableTreeSitter, getTreeSitter, } from "./utils/tree-sitter.js";
export { charWidth, stringWidth, iterGraphemes } from "./core/unicode.js";
export { WebRenderer } from "./core/web-renderer.js";
export { ResizeObserver, notifyResizeObservers, } from "./core/resize-observer.js";
export { StormSSHServer } from "./ssh/index.js";
export { TestInputManager, MockInputManager, renderForTest, expectOutput, createSnapshot, compareSnapshot, clearSnapshots, saveSnapshot, compareFileSnapshot, saveSvgSnapshot, compareSvgSnapshot, createStormMatchers, renderToSvg, } from "./testing/index.js";
// Animation & timing hooks
export { useTick } from "./hooks/useTick.js";
export { usePhaseTimer } from "./hooks/usePhaseTimer.js";
export { useProfiler } from "./hooks/useProfiler.js";
